Add Road to Component Model 1.0 blog (#147)
bytecodealliance/bytecodealliance.org: _posts/2026-04-21-the-road-to-component-model-1-0.md

---
title: "The Road to Component Model 1.0"
author: "Eric Gregory"

> date: "2026-04-21"

Reviewer: Let's update this once we know a release date (hopefully Real Soon Now!)

github_name: "ericgregory"
excerpt_separator: <!--end_excerpt-->
---
WASI P3 is almost here, bringing native async support to the **WebAssembly System Interface (WASI)** and **Component Model**. In this post, we're looking to the *next* big milestone: a stable, formally specified Component Model 1.0.

> At February's Bytecode Alliance Plumbers Summit, Luke Wagner and Alex Crichton gave a preview of what the path to a stable 1.0 actually looks like. Last month at [Wasm I/O 2026 in Barcelona](https://wasm.io/), Wagner [expanded on that vision](https://youtu.be/qq0Auw01tH8?si=4Zb2r52SNsAsUtkQ). So let's take a look at where the Component Model is heading.

Reviewer: ```suggestion
At February's Bytecode Alliance Plumbers Summit, Luke Wagner and Alex Crichton gave a preview of what the path to a stable 1.0 actually looks like. At [Wasm I/O 2026 in Barcelona](https://wasm.io/) in March, Wagner [expanded on that vision](https://youtu.be/qq0Auw01tH8?si=4Zb2r52SNsAsUtkQ). So let's take a look at where the Component Model is heading.
```

> At February's Bytecode Alliance Plumbers Summit, Luke Wagner and Alex Crichton gave a preview of what the path to a stable 1.0 actually looks like. Last month at [Wasm I/O 2026 in Barcelona](https://wasm.io/), Wagner [expanded on that vision](https://youtu.be/qq0Auw01tH8?si=4Zb2r52SNsAsUtkQ). So let's take a look at where the Component Model is heading.

Reviewer: Here and everywhere else below, I think it'd fit the tone of our blog a bit better to use first instead of last names

WASI interfaces are consumed through the Component Model; together they form the composable, portable foundation of the Wasm ecosystem. (It's worth noting that Component Model 1.0 and WASI 1.0 are related but distinct milestones: WASI 1.0 will follow from and depend on the Component Model reaching 1.0 first.)

At the Plumbers Summit, Wagner described the relationship between the Component Model and WASI as analogous to a microkernel architecture: the Component Model is the always-present microkernel, providing foundational primitives that run across any host. WASI layers on top like OS services (e.g., networking, storage, graphics) that run as processes on top of the microkernel and may or may not be present on a given device. A browser, for instance, has very strong opinions about what I/O APIs exist; WASI interfaces run there via polyfill. But the Component Model itself can be implemented natively in the browser alongside core WebAssembly, since it only provides computational primitives, not I/O.

> In practical terms, the Component Model is already stable. P1 modules still work. P2 components still work. The team has been maintaining this stability since P1 using semantic versioning, side-by-side implementations, and Wasm-to-Wasm adapters. That story continues past 1.0. As Wagner put it: "We can start using this stuff now." But *getting* to Component Model 1.0 involves critical work across several important areas.

Reviewer: Slightly reworded to also include WASI, because otherwise the mention of (pre-CM) P1 feels a bit off. Plus, it seems important to highlight that all this applies to WASI as well.

```suggestion
In practical terms, both WASI and the Component Model are already heavily used in production. Despite ongoing evolution of both the binary format and WASI APIs, platform providers and embedders can give strong backwards-compatibility guarantees. P1 modules still work. P2 components still work. The team has been maintaining this stability since P1 using semantic versioning, side-by-side implementations, and Wasm-to-Wasm adapters. That story continues past 1.0. As Wagner put it: "We can start using this stuff now." But *getting* to Component Model 1.0 involves critical work across several important areas.
```

Reviewer: Note that I didn't change the mention of `Wagner` in here, since that suggestion from above might be rejected.

## Five areas of work

Wagner frames the path to 1.0 around five areas of work:

### 1. ABI improvements

> The Component Model's current Application Binary Interface (ABI), the calling convention that governs how components pass data to each other, relies on a function conventionally called `cabi_realloc`. When a callee returns a value like a list of strings, the host calls `cabi_realloc` to allocate memory for each element, then copies everything before returning to the caller. This works, but experience has surfaced real friction: heap fragmentation over time, difficulty handling large allocation failures gracefully, one host-to-guest call per value in a list, and trouble using custom memory allocators.

Reviewer: ```suggestion
The Component Model's current Application Binary Interface (ABI), the calling convention that governs how components pass data to each other, relies on a function conventionally called `cabi_realloc`, which all components are required to export. When a callee returns a value like a list of strings, the host calls the calling component's `cabi_realloc` to allocate memory for each element, then copies everything before returning to the caller. This works, but experience has surfaced real friction: heap fragmentation over time, difficulty handling large allocation failures gracefully, one host-to-guest call per value in a list, and trouble using custom memory allocators.
```

![Two sequence diagrams comparing the eager and lazy ABIs. In the eager diagram, the caller invokes the callee, and the callee makes repeated round-trip calls to cabi_realloc to allocate memory for each returned value before control returns to the caller. In the lazy diagram, the callee returns immediately with opaque handles; the caller then calls statically-known ABI built-ins that are inlined directly into the caller rather than crossing a component boundary.](../images/lazy-and-eager.png)

The transition is designed to be non-breaking: the lazy ABI ships as a new opt-in option in a 0.3.x minor release, and producer toolchains adopt it when ready. When 1.0 arrives and adoption is complete, the lazy ABI becomes the default and the option goes away. An adapter tool will handle converting eager components to lazy.

Bundled into the same transition are multivalue returns (using Wasm multivalue for return types at the C ABI level) and an error context value in every result's error case, so there's always a standard way for hosts to attach backtraces and debug information.

> There is one significant dependency: LLVM doesn't yet support multivalue at the C ABI level. Getting that upstream may be the longest lead time in this piece of work.

Reviewer: Hot off the presses:

```suggestion
There is one significant dependency: LLVM doesn't yet support multivalue at the C ABI level. A [proposal for the precise ABI](https://github.com/WebAssembly/tool-conventions/pull/268) exists, but getting that upstream may be the longest lead time in this piece of work.
```

There's also a separate GC ABI option planned, enabling components to pass values via Wasm GC-allocated memory instead of linear memory, and avoiding copies through linear memory for GC-based languages. Nick Fitzgerald has a pre-proposal on this in [Component Model issue 525](https://github.com/WebAssembly/component-model/issues/525).

> Sitting alongside the ABI work is a related performance goal: zero overhead on synchronous calls between components. The current implementation manages async task infrastructure across component boundaries via a host call, which (per Nick Fitzgerald's Plumbers Summit measurements) adds roughly 3.5x overhead even on purely synchronous call paths. Some of this has already been addressed at the spec level (the recursion check was removed during P3 development). The remaining work is [planned](https://github.com/bytecodealliance/wasmtime/issues/12311) for after WASI P3 ships, refactoring task state so synchronous adapters allocate a lightweight task on the stack that [Cranelift](https://github.com/bytecodealliance/wasmtime/tree/main/cranelift) (Wasmtime's optimizing code generation backend) can largely optimize away. The goal is to restore the performance profile that synchronous adapters had before component model async support was added.

Reviewer: ```suggestion
Sitting alongside the ABI work is a related performance goal: zero overhead on synchronous calls between components. The current implementation manages async task infrastructure across component boundaries via a host call, which (per Nick Fitzgerald's Plumbers Summit measurements) adds roughly 3.5x overhead even on purely synchronous call paths. Some of this has already been addressed at the spec level (the recursion check was removed during P3 development). The remaining work is [planned](https://github.com/bytecodealliance/wasmtime/issues/12311) for after WASI P3 ships, refactoring task state so synchronous adapters allocate a lightweight task on the stack that compilers like [Cranelift](https://github.com/bytecodealliance/wasmtime/tree/main/cranelift) (Wasmtime's optimizing code generation backend) can largely optimize away. The goal is to restore the performance profile that synchronous adapters had before component model async support was added.
```

The strategy for earning that support borrows a page from [WebAssembly's own history](https://bytecodealliance.org/articles/ten-years-of-webassembly-a-retrospective). When asm.js shipped, it contained a no-op `"use asm"` string that browser engines could detect in their feature usage telemetry, building a data-driven case for optimization and eventually WebAssembly itself. `jco` is doing the same: its generated JavaScript glue now emits a `"use components"` identifier (renamed from `"use jco"` in jco 1.16.8). As `jco`-transpiled components accumulate real-world usage, that data becomes evidence for native implementation.

Both Mozilla and Chrome are paying attention. The Mozilla team presented performance results at a recent in-person WebAssembly CG meeting. The Chrome V8 team opened an issue for evaluating component model implementation. These aren't commitments, but they're meaningful signals. For folks working in the component ecosystem, now is a great time to use `jco`-transpiled components in browser contexts.

### 3. Making the Component Model easier to implement

> Getting native browser support requires a manageable implementation burden. The plan here is a pair of C ABIs, one for guests and one for hosts, both expressed as header files generated by `wit-bindgen`.

Reviewer: I added language here on what I think is a very important additional aspect of making implementation easier: the planned simplification of the spec.

```suggestion
Getting native browser support requires a manageable implementation burden. The plan here has two key components:

The first one is to simplify the specification itself. The Component Model 1.0 spec will only include the "good parts" of the P3 spec, instead of being a superset of what exists right now. Most importantly, the ABI change described above means that implementers aiming for 1.0, but not P3, support don't have to support the current, `cabi_realloc` based calling convention. As is the case for P1 and P2 support in various production environments today, existing P3 components will still be supported with the help of tools and adapters.

The second one is a pair of C ABIs to ease implementation, one for guests and one for hosts, both expressed as header files generated by `wit-bindgen`.
```

- **Map type**: For producing and consuming idiomatic objects, dictionaries, and associative arrays across languages. (Yordis Prieto is already working on this!)

- **Runtime instantiation**: The ability to instantiate components at runtime, within the Component Model. This is useful for lazy code loading, spawning subcommands, and giving different components different lifetimes. Browsers already support something more powerful in their JavaScript API; this would bring some of this functionality to the Component Model so that producer toolchains could address these use cases in a way that works both inside *and* outside the browser.

Not all of these will land before 1.0: some may come in 1.x releases. Sequencing depends on use cases and available resources.

## Cooperative threads and stream splicing

Sy Brand presented a deep dive on cooperative threads at the Plumbers Summit, and it's worth noting here that they're implemented at the Component Model level, below the WASI layer. A Rust component using `std::thread::spawn` gets cooperative thread support when it lands, with nothing special needed in WIT.

> The implementation is in progress: `wasi-libc` pthreads support is largely done, LLVM patches are under review, and Wasmtime has a functional implementation under a feature flag. Cooperative threads support will ship in a follow-up WASI P3 minor release. These advancements also pave the road for shared-everything threads later, since many of the heaviest toolchain lifts for cooperative threads carry over directly.

Reviewer: ```suggestion
The implementation is in progress: `wasi-libc` pthreads support is largely done, LLVM patches have just landed and will be included in the next release, and Wasmtime has a functional implementation under a feature flag. Cooperative threads support will ship in a follow-up WASI P3 minor release. These advancements also pave the road for shared-everything threads later, since many of the heaviest toolchain lifts for cooperative threads carry over directly.
```

> Also shipping in an early WASI P3 follow-up release is **stream splicing**, the ability to splice one stream directly into another without an intermediate copy. Stream splicing is important for streaming performance, since unnecessary copying through an intermediate buffer is costly on hot paths, and was simply too close to the wire to make WASI P3 itself.

Reviewer: ```suggestion
Also shipping in an early WASI P3 follow-up release will be **stream splicing**, the ability to splice one stream directly into another without an intermediate copy. Stream splicing is important for streaming performance, since unnecessary copying through an intermediate buffer is costly on hot paths, and was simply too close to the wire to make WASI P3 itself.
```

<!--end_excerpt-->

## A quick refresher on the Component Model and WASI

Before we dive into the future of the Component Model, we should take a moment to review what the specification does, as well as its relationship to WASI.

* The [**Component Model**](https://github.com/WebAssembly/component-model) is the specification layered atop core WebAssembly that defines how Wasm binaries bundle, link, and communicate: it specifies the type system, the binary format, the interface definition language, and the calling conventions for passing typed values across component isolation boundaries. 

* The [**WebAssembly System Interface (WASI)**](https://github.com/WebAssembly/WASI), in turn, is a standardized set of APIs for accessing system resources like files, sockets, clocks, randomness, and more in a portable, capability-secure way. 

WASI interfaces are consumed through the Component Model; together they form the composable, portable foundation of the Wasm ecosystem. (It's worth noting that Component Model 1.0 and WASI 1.0 are related but distinct milestones: WASI 1.0 will follow from and depend on the Component Model reaching 1.0 first.)

> At the Plumbers Summit, Wagner described the relationship between the Component Model and WASI as analogous to a microkernel architecture: the Component Model is the always-present microkernel, providing foundational primitives that run across any host. WASI layers on top like OS processes (e.g., networking, storage, graphics) that may or may not be present on a given device. A browser, for instance, has very strong opinions about what I/O APIs exist; WASI interfaces run there via polyfill. But the Component Model itself can be implemented natively in the browser alongside core WebAssembly.

Reviewer 2: ```suggestion
At the Plumbers Summit, Wagner described the relationship between the Component Model and WASI as analogous to a microkernel architecture: the Component Model is the always-present microkernel, providing foundational primitives that run across any host. WASI layers on top like OS services (e.g., networking, storage, graphics) that run as processes on top of the microkernel and may or may not be present on a given device. A browser, for instance, has very strong opinions about what I/O APIs exist; WASI interfaces run there via polyfill. But the Component Model itself can be implemented natively in the browser alongside core WebAssembly, since it only provides computational primitives, not I/O.
```

> In practical terms, the Component Model is already stable. P1 components still work. P2 components still work. The team has been maintaining this stability since P1 using semantic versioning, side-by-side implementations, and Wasm-to-Wasm adapters. That story continues past 1.0. As Wagner put it: "We can start using this stuff now." But *getting* to Component Model 1.0 involves critical work across several important areas.

Reviewer 2: ```suggestion
In practical terms, the Component Model is already stable. P1 modules still work. P2 components still work. The team has been maintaining this stability since P1 using semantic versioning, side-by-side implementations, and Wasm-to-Wasm adapters. That story continues past 1.0. As Wagner put it: "We can start using this stuff now." But *getting* to Component Model 1.0 involves critical work across several important areas.
```

## Five areas of work

Wagner frames the path to 1.0 around five areas of work:

### 1. ABI improvements

> The Component Model's current Application Binary Interface (ABI), the calling convention that governs how components pass data to each other, relies on a function called `cabi_realloc`. When a callee returns a value like a list of strings, the host calls `cabi_realloc` to allocate memory for each element, then copies everything before returning to the caller. This works, but experience has surfaced real friction: heap fragmentation over time, difficulty handling large allocation failures gracefully, and one host-to-guest call per value in a list.

Reviewer 2: ```suggestion
The Component Model's current Application Binary Interface (ABI), the calling convention that governs how components pass data to each other, relies on a function conventionally called `cabi_realloc`. When a callee returns a value like a list of strings, the host calls `cabi_realloc` to allocate memory for each element, then copies everything before returning to the caller. This works, but experience has surfaced real friction: heap fragmentation over time, difficulty handling large allocation failures gracefully, one host-to-guest call per value in a list, and trouble using custom memory allocators.
```

![Two sequence diagrams comparing the eager and lazy ABIs. In the eager diagram, the caller invokes the callee, and the callee makes repeated round-trip calls to cabi_realloc to allocate memory for each returned value before control returns to the caller. In the lazy diagram, the callee returns immediately with opaque handles; the caller then calls statically-known ABI built-ins that are inlined directly into the caller rather than crossing a component boundary.](../images/lazy-and-eager.png)

The transition is designed to be non-breaking: the lazy ABI ships as a new opt-in option in a 0.3.x minor release, and producer toolchains adopt it when ready. When 1.0 arrives and adoption is complete, the lazy ABI becomes the default and the option goes away. An adapter tool will handle converting eager components to lazy.

Bundled into the same transition are multivalue returns (using Wasm multivalue for return types at the C ABI level) and an error context value in every result's error case, so there's always a standard way for hosts to attach backtraces and debug information.

> A related but distinct performance goal sits alongside the ABI work: sync-to-sync component-to-component calls with zero overhead. The current spec includes async state management flags that, according to Nick Fitzgerald's Plumbers Summit presentation, cause roughly 3.5x overhead on synchronous calls. These flags exist to handle reentrance and task switching in the async runtime, but they add cost even for purely synchronous call paths. Some of this has already been addressed at the spec level (the recursion check was removed during P3 development), and active work is underway in [Wasmtime](https://github.com/bytecodealliance/wasmtime) to move remaining task state into inline trampolines that [Cranelift](https://github.com/bytecodealliance/wasmtime/tree/main/cranelift) (Wasmtime's optimizing code generation backend) can optimize away.

Reviewer 2: It might be helpful to get an issue URL from Alex to link to from "active work is underway" (and maybe ask whether we should change "underway" to "planned"...).

Reviewer: Even better: the fix for this [just landed](https://github.com/bytecodealliance/wasmtime/pull/13194). It relies on inlining being enabled. That's already possible to do manually, with #13214 enabling it by default. Chances are that will also land before this post is published.

Author: I updated the language to reframe this as a solved problem and refer to the work in the past tense (also linking the issues), so we could plan to publish this after inlining by default lands.

Reviewer 2: @tschneidereit Check me on this, but, iiuc, while `context.{get,set}` inlining was *one* key perf improvement, another is the inlining and "lazifying" of the task structures.  IIUC, until this is done, we still have a host call per cross-component call (even sync).  But happy to be corrected!

Reviewer 3: Luke is correct, yeah, the 3.5x overhead Nick presented about is still present and hasn't been fixed yet. A related, but mostly orthogonal, optimization Till mentioned has landed but is unrelated to this paragraph.

Reviewer 2: Is there a good issue we could link to from "planned" (just so folks can follow and, hopefully later, see that it's fixed)?

Reviewer 3: https://github.com/bytecodealliance/wasmtime/issues/12311

Author: I've updated the blog to reflect this conversation -- just let me know if anything needs to be adjusted there.

Reviewer 2: lgtm, thanks!

There is one significant dependency: LLVM doesn't yet support multivalue at the C ABI level. Getting that upstream may be the longest lead time in this piece of work.

There's also a separate GC ABI option planned, enabling components to pass values via Wasm GC-allocated memory instead of linear memory, and avoiding copies through linear memory for GC-based languages. Nick Fitzgerald has a pre-proposal on this in [Component Model issue 525](https://github.com/WebAssembly/component-model/issues/525).

### 2. The browser path

The Component Model can't formally reach 1.0 without native implementation in at least two browser engines. The groundwork for browser implementations is being laid today: `jco`'s `transpile` command already converts any component into equivalent core Wasm and JavaScript glue, making components runnable in any browser without native support. That already works, but native support matters for two reasons:

> - **Performance**: Experiments by Ryan Hunt at Mozilla show that a DOM mutation-heavy Wasm VDOM reconciliation loop can get close to a 2x speedup from direct Wasm-to-browser API calls, bypassing the JavaScript glue layer. Real-world gains will vary by workload, but the ceiling is meaningful.

Reviewer 2: ```suggestion
- **Performance**: [Experiments](https://hacks.mozilla.org/wp-content/uploads/2026/02/Screenshot-2026-02-25-at-2.22.23-PM-1536x1018.png) by Ryan Hunt at Mozilla show that a DOM mutation-heavy Wasm VDOM reconciliation loop can get close to a 2x speedup from direct Wasm-to-browser API calls, bypassing the JavaScript glue layer. Real-world gains will vary by workload, but the ceiling is meaningful.
```

- **Reach**: Native Component Model support in browsers creates strong incentives for upstream Wasm support across languages, packages, and frameworks, which in turn feeds back into a better developer experience across all platforms where Wasm runs.

The strategy for earning that support borrows a page from [WebAssembly's own history](https://bytecodealliance.org/articles/ten-years-of-webassembly-a-retrospective). When asm.js shipped, it contained a no-op `"use asm"` string that browser engines could detect in their feature usage telemetry, building a data-driven case for optimization and eventually WebAssembly itself. `jco` is doing the same: its generated JavaScript glue now emits a `"use components"` identifier (renamed from `"use jco"` in jco 1.16.8). As `jco`-transpiled components accumulate real-world usage, that data becomes evidence for native implementation.

> Both Mozilla and Chrome are paying attention. The Mozilla team presented performance results at a recent in-person WebAssembly CG meeting. The Chrome V8 team opened an issue for evaluating component model implementation. These aren't commitments, but they're meaningful signals. For folks working in the component ecosystem, now is a great time to use `jco`-transpiled components in browser contexts.

Reviewer 2: Could you reword the Mozilla sentence to mention and link to their Hacks blog post (https://hacks.mozilla.org/2026/02/making-webassembly-a-first-class-language-on-the-web/) and have the V8 sentence link "issue" to https://issues.chromium.org/issues/474661098?

### 5. Closing expressivity gaps

[WebAssembly Interface Type (WIT)](https://component-model.bytecodealliance.org/design/wit.html) serves as the Component Model's interface definition language, describing the typed APIs a component imports and exports.

The path to 1.0 includes work on several WIT features, based on gaps identified in practice:

- **Stream v2**: The current 0.3.0 HTTP body type requires a complex function signature to represent all the concurrent things that can happen while consuming a request or response: a byte stream, a final result (trailers or error), and a back-channel for the consumer to signal the producer. A generalized stream type would make this pattern readable and reusable across any streaming protocol with state transitions.

> 

Reviewer 2: ```suggestion
```
I ended up cutting this one from the final Wasm I/O pres (it's a bit complex, a little below-the-fold, and it doesn't have a great write-up anywhere to link to).

- **Optional imports**: Lets a component declare that it only optionally requires a capability. Standard libraries could compile against a minimal host world without forcing every component to unconditionally import files, sockets, and GPU capabilities. This pairs well with the guest C-ABI.

- **Callbacks**: Needed to express DOM APIs like `addEventListener` in WIT.

> - **Resource subtyping**: Lets a resource type (like a DOM Node) extend another (like EventTarget), enabling method inheritance without virtual dispatch or complex inheritance semantics.

Reviewer 2: ```suggestion
- **Resource subtyping**: Lets a resource type (like a DOM Node) extend another (like EventTarget), enabling method reuse without virtual dispatch or complex inheritance semantics.
```

> - **Function subtyping**: Allows WIT interfaces to add parameters, record fields, and variant cases without breaking changes, something that's currently painful in practice.

Reviewer 2: ```suggestion
- **Function subtyping**: Allows WIT interfaces to add parameters, record fields, and variant cases without having to add new functions or declare a breaking change, something that's currently painful in practice.
```

> - **Parameterized import/export names**: Lets components encode metadata in their import names (for example, which key-value store backend they expect) enabling deployment-time configuration checking rather than runtime surprises.

Reviewer 2: ```suggestion
- **Enhanced import/export names**: Lets components import multiple of the same WIT interface with different arbitrary identifiers, enabling deployment-time configuration checking rather than runtime surprises.
```

- **Getters and setters**: Syntactic sugar for more idiomatic bindings in languages that expect property access patterns. Like constructors and methods, these would be sugar for regular functions with special names that binding generators recognize.

- **Map type**: For producing and consuming idiomatic objects, dictionaries, and associative arrays across languages. (Yordis Prieto is already working on this!)

> - **Runtime instantiation**: The ability to instantiate components at runtime, within the Component Model. This is useful for lazy code loading, spawning subcommands, and giving different components different lifetimes. Browsers already support something more powerful in their JavaScript API; this would bring equivalent functionality to the Component Model.

Reviewer 2: ```suggestion
- **Runtime instantiation**: The ability to instantiate components at runtime, within the Component Model. This is useful for lazy code loading, spawning subcommands, and giving different components different lifetimes. Browsers already support something more powerful in their JavaScript API; this would bring some of this functionality to the Component Model so that producer toolchains could address these use cases in a way that works both inside *and* outside the browser.
```
